bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v3/product/ProductImpl.java

```diff
@@ -0,0 +1,91 @@
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.commerce.core.components.internal.models.v3.product;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.models.annotations.Model;
+
+import com.adobe.cq.commerce.core.components.internal.models.v1.product.VariantAttributeImpl;
+import com.adobe.cq.commerce.core.components.internal.models.v1.product.VariantValueImpl;
+import com.adobe.cq.commerce.core.components.models.product.Product;
+import com.adobe.cq.commerce.core.components.models.product.Variant;
+import com.adobe.cq.commerce.core.components.models.product.VariantAttribute;
+import com.adobe.cq.commerce.core.components.models.product.VariantValue;
+import com.adobe.cq.commerce.magento.graphql.ConfigurableAttributeOption;
+import com.adobe.cq.commerce.magento.graphql.ConfigurableProductOptions;
+import com.adobe.cq.commerce.magento.graphql.ConfigurableProductOptionsValues;
+import com.adobe.cq.commerce.magento.graphql.ConfigurableVariant;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = Product.class, resourceType = ProductImpl.RESOURCE_TYPE)
+public class ProductImpl extends com.adobe.cq.commerce.core.components.internal.models.v2.product.ProductImpl
+    implements Product {
+
+    public static final String RESOURCE_TYPE = "core/cif/components/commerce/product/v3/product";
+
+    @PostConstruct
+    protected void initModel() {
+        super.initModel();
+        if (productRetriever != null) {
+            productRetriever.extendProductQueryWith(p -> p.onConfigurableProduct(cp -> cp
+                .configurableOptions(o -> o
+                    .values(v -> v.uid()))
+                .variants(v -> v
+                    .attributes(a -> a.uid()))));
+        }
+    }
+
+    @Override
+    protected VariantValue mapVariantValue(ConfigurableProductOptionsValues value) {
+        VariantValueImpl variantValue = new VariantValueImpl();
+        variantValue.setId(value.getValueIndex());
+        variantValue.setUid(value.getUid().toString());
+        variantValue.setLabel(value.getLabel());
+
+        return variantValue;
+    }
+
+    @Override
+    protected VariantAttribute mapVariantAttribute(ConfigurableProductOptions option) {
+        // Get list of values
+        List<VariantValue> values = option.getValues().parallelStream().map(this::mapVariantValue)
+            .collect(Collectors.toList());
+
+        // Create attribute map
+        VariantAttributeImpl attribute = new VariantAttributeImpl();
+        attribute.setLabel(option.getLabel());
+        attribute.setId(option.getAttributeCode());
+        attribute.setValues(values);
+
+        return attribute;
+    }
+
+    @Override
+    protected Variant mapVariant(ConfigurableVariant variant) {
+        Variant mappedVariant = super.mapVariant(variant);
```

Comment:
Maybe it would be better to return the `VariantImpl` from the `mapVariant()` call and add a method to it to mutate the internal state of the `variantAttributesUid`.  Ideally the collections/maps returned by a data object are immutable.

Reply:
Agreed. But I was following the same pattern applied for `getVariantAttributes`.